src/System/Terminal/Widgets/TextInput.hs

```diff
@@ -32,30 +33,44 @@ instance {-# OVERLAPPING #-} Lens.HasField "cursor" TextInput TextInput Cursor C
             RopeZipper.Position{posLine = fromIntegral row, posColumn = fromIntegral col}
 
 instance Widget TextInput where
-    handleEvent (KeyEvent BackspaceKey []) = #value %~ RopeZipper.deleteBefore
-    handleEvent (KeyEvent DeleteKey []) = #value %~ RopeZipper.deleteAfter
-    handleEvent (KeyEvent (CharKey k) []) = #value %~ RopeZipper.insertText (Text.singleton k)
-    handleEvent (KeyEvent (ArrowKey Leftwards) []) = #value %~ RopeZipper.moveBackward
-    handleEvent (KeyEvent (ArrowKey Rightwards) []) = #value %~ RopeZipper.moveForward
-    handleEvent (KeyEvent EnterKey []) = filtered (.multiline) . #value %~ RopeZipper.insertText "\n"
-    handleEvent (KeyEvent (ArrowKey Upwards) []) = filtered (.multiline) . #value %~ RopeZipper.moveUp
-    handleEvent (KeyEvent (ArrowKey Downwards) []) = filtered (.multiline) . #value %~ RopeZipper.moveDown
-    handleEvent _ = id
+    handleEvent (KeyEvent BackspaceKey []) t@(TextInput{value}) =
+        t{value = RopeZipper.deleteBefore value}
+    handleEvent (KeyEvent DeleteKey []) t@(TextInput{value}) =
+        t{value = RopeZipper.deleteAfter value}
+    handleEvent (KeyEvent (CharKey k) []) t@(TextInput{value}) =
+        t{value = RopeZipper.insertText (Text.singleton k) value}
+    handleEvent (KeyEvent (ArrowKey Leftwards) []) t@(TextInput{value}) =
+        t{value = RopeZipper.moveBackward value}
+    handleEvent (KeyEvent (ArrowKey Rightwards) []) t@(TextInput{value}) =
+        t{value = RopeZipper.moveForward value}
+    handleEvent (KeyEvent EnterKey []) t@(TextInput{multiline, value}) =
+        if multiline
+            then t{value = RopeZipper.insertText "\n" value}
+            else t
+    handleEvent (KeyEvent (ArrowKey Upwards) []) t@(TextInput{multiline, value}) =
+        if multiline
+            then t{value = RopeZipper.moveUp value}
+            else t
+    handleEvent (KeyEvent (ArrowKey Downwards) []) t@(TextInput{multiline, value}) =
+        if multiline
+            then t{value = RopeZipper.moveDown value}
+            else t
```

Comment:
Let's simplify these with `| multiline = t{...}` and let the wildcard do the `t` case.